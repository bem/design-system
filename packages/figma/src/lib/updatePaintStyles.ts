import { isTokenPrimitiveValue, isTokensGroup, isTokenValue, TokensGroup } from './TokensData'
import { hexToFigmaRGB, webRGBToFigmaRGB } from '@figma-plugin/helpers'
import { cssRGBtoWebRgb, isHex, isRgb } from './colors'

export const toFigmaRGB = (value: string) => {
  if (isHex(value)) {
    return hexToFigmaRGB(value)
  } else if (isRgb(value)) {
    return webRGBToFigmaRGB(cssRGBtoWebRgb(value))
  }
  return hexToFigmaRGB('#000')
}

export function mapTokensGroupToFigmaGroup(tokensGroup: TokensGroup) {
  const styles: TokensGroup = {}

  const queue = Object.entries(tokensGroup)

  while (queue.length > 0) {
    const entry = queue.shift()!

    const [key, value] = entry
    if (isTokenPrimitiveValue(value) || isTokenValue(value)) {
      styles[key] = value
    } else if (isTokensGroup(value)) {
      Object.entries(value).forEach(([itemKey, itemValue]) => {
        queue.push([`${key}/${itemKey}`, itemValue])
      })
    }
  }
  return styles
}

export const updatePaintStyle = (name: string, color: ReturnType<typeof toFigmaRGB>, description?: string) => {
  return Object.assign(figma.createPaintStyle(), {
    name,
    paints: [
      {
        type: 'SOLID',
        color,
        opacity: 1,
      },
    ],
    description,
  })
}

export const updatePaintStyles = (colors: TokensGroup) => {
  const styles = mapTokensGroupToFigmaGroup(colors)

  for (const [name, value] of Object.entries(styles)) {
    if (isTokenPrimitiveValue(value)) {
      updatePaintStyle(name, toFigmaRGB(value as string))
    } else if (isTokenValue(value)) {
      updatePaintStyle(name, toFigmaRGB(value.value as string), value.description)
    }
  }
}
