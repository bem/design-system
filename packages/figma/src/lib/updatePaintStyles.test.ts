import { mapTokensGroupToFigmaGroup, toFigmaRGB, updatePaintStyle, updatePaintStyles } from './updatePaintStyles'
import * as B from './updatePaintStyles'

// Mock global the figma namespace
Object.defineProperty(global, 'figma', {
  value: {
    createPaintStyle() {
      return {}
    },
  },
  writable: true,
})

describe('toFigmaRGB', () => {
  test('return value as figma PaintStyle', () => {
    expect(toFigmaRGB('#fff')).toMatchSnapshot()
    expect(toFigmaRGB('rgba(0, 0, 0)')).toMatchSnapshot()
    expect(toFigmaRGB('rgba(0, 0, 0, 0.8)')).toMatchSnapshot()
  })

  test('return default value as figma PaintStyle', () => {
    expect(toFigmaRGB('hsl(0, 0, 0)')).toMatchSnapshot()
  })
})

describe('updatePaintStyle', () => {
  test('createPaintStyle have been called and return object', () => {
    const spy = jest.spyOn(figma, 'createPaintStyle')
    const paintStyle = updatePaintStyle('Red', toFigmaRGB('#f00'), 'Basic color')
    expect(spy).toHaveBeenCalledTimes(1)
    expect(paintStyle).toMatchSnapshot()
    spy.mockRestore()
  })
})

describe('updatePaintStyles', () => {
  test('call updatePaintStyle', () => {
    const mock = jest.spyOn(B, 'updatePaintStyle')
    updatePaintStyles({
      Red: {
        '100': '#f00',
        '200': {
          value: '#f10',
          description: 'Basic color',
        },
      },
      Action: {
        Base: {
          value: '#0ff',
          description: 'Basic color',
        },
      },
    })

    expect(mock).toHaveBeenNthCalledWith(1, 'Red/100', toFigmaRGB('#f00'))
    expect(mock).toHaveBeenNthCalledWith(2, 'Red/200', toFigmaRGB('#f10'), 'Basic color')
    expect(mock).toHaveBeenNthCalledWith(3, 'Action/Base', toFigmaRGB('#0ff'), 'Basic color')
    mock.mockRestore()
  })
})

describe('mapTokensGroupToFigmaGroup', () => {
  test('return correct mapping', () => {
    expect(
      mapTokensGroupToFigmaGroup({
        Red: {
          '100': '#f00',
          '200': {
            value: '#f10',
            description: 'Basic color',
          },
        },
        Action: {
          Base: {
            value: '#0ff',
            description: 'Basic color',
          },
        },
      }),
    ).toMatchSnapshot()
  })
})
