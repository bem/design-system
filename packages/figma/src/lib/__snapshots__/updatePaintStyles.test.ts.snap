// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapTokensGroupToFigmaGroup return correct mapping 1`] = `
Object {
  "Action/Base": Object {
    "description": "Basic color",
    "value": "#0ff",
  },
  "Red/100": "#f00",
  "Red/200": Object {
    "description": "Basic color",
    "value": "#f10",
  },
}
`;

exports[`toFigmaRGB return default value as figma PaintStyle 1`] = `
Object {
  "b": 0,
  "g": 0,
  "r": 0,
}
`;

exports[`toFigmaRGB return value as figma PaintStyle 1`] = `
Object {
  "b": 1,
  "g": 1,
  "r": 1,
}
`;

exports[`toFigmaRGB return value as figma PaintStyle 2`] = `
Object {
  "b": 0,
  "g": 0,
  "r": 0,
}
`;

exports[`toFigmaRGB return value as figma PaintStyle 3`] = `
Object {
  "a": 0.8,
  "b": 0,
  "g": 0,
  "r": 0,
}
`;

exports[`updatePaintStyle createPaintStyle have been called and return object 1`] = `
Object {
  "description": "Basic color",
  "name": "Red",
  "paints": Array [
    Object {
      "color": Object {
        "b": 0,
        "g": 0,
        "r": 1,
      },
      "opacity": 1,
      "type": "SOLID",
    },
  ],
}
`;
